### PR TITLE
Unity Embedding API

### PR DESCRIPTION
**Pull request by joncham**
Heavily refactored from initial prototype repo commit: 5eac4186181f9b54a82fa49a30a5baf3b138c7bd

This enables the mono embedding API from within the coreclr dynamic library. 

- Initial commits are atomic pieces of functionality to CoreCLR itself which we depend on. These should be reviewed and potentially upstreamed in some form.
- Final commit is bulk commit of embedding API work.
- It is functional on amd64 arch of Windows, Mac, and Linux.
- Non functionality on i386 Windows due to our use of the bump allocation unitygc
- Embedded tests run on Windows and Mac. Linux had issues running `dotnet` command which may be issue with image.

### CODE REVIEW

File: src/coreclr/vm/classlayoutinfo.cpp [hunk 1]
```diff
@@ -55,9 +55,14 @@ namespace
                 pfwalk->m_placement.m_offset = ulOffset;
                 pfwalk->m_sequence = (ULONG)-1;
 
+#if FEATURE_UNITY_EMBEDDING_INTERFACE
```
Comment (scott-ferguson-unity) [inline src/coreclr/vm/classlayoutinfo.cpp:58]: This bug may have been fixed: See https://github.com/dotnet/runtime/issues/53542 & https://github.com/dotnet/runtime/pull/54235

Comment (joncham) [reply src/coreclr/vm/classlayoutinfo.cpp:58]: Nice. Thanks for finding this @scott-ferguson-unity  !

File: .yamato/build_windows.yml [hunk 2]
```diff
@@ -15,6 +15,12 @@ commands:
   - powershell .yamato\scripts\download_7z.ps1
   - artifacts\7za-win-x64\7za.exe a artifacts\unity\dotnet-runtime-unity-win-x86.7z .\artifacts\bin\microsoft.netcore.app.runtime.win-x86\Release\runtimes\win-x86
```
Comment (UnityAlex) [inline .yamato/build_windows.yml]:
```suggestion
  - artifacts\7za-win-x64\7za.exe a artifacts\unity\dotnet-runtime-unity-win-x86.7z .\artifacts\bin\microsoft.netcore.app.runtime.win-x86\Release\runtimes\win-x86\*
```
Missed that in my first pass.

Comment (joncham) [reply .yamato/build_windows.yml]: Done

File: src/coreclr/vm/classlayoutinfo.cpp (hunk 1 above)
Comment (joncham) [reply src/coreclr/vm/classlayoutinfo.cpp]: I tried removing this and the issue seemed to remain.

File: src/coreclr/inc/crsttypes.h [hunk 3]
```diff
@@ -84,57 +84,59 @@ enum CrstType
...
-    CrstUMEntryThunkFreeListLock = 113,
-    CrstUniqueStack = 114,
-    CrstUnresolvedClassLock = 115,
-    CrstUnwindInfoTableLock = 116,
-    CrstVSDIndirectionCellLock = 117,
-    CrstWrapperTemplate = 118,
-    kNumberOfCrstTypes = 119
+    CrstMonoHandles = 69,
+    CrstMonoICalls = 70,
+    CrstMulticoreJitHash = 71,
```
Comment (UnityAlex) [inline src/coreclr/inc/crsttypes.h:89]: Will changing the value of `CrstMulticoreJitHash` (and maybe others) cause adverse affects? Would they accept this upstream?

Comment (jkotas) [reply src/coreclr/inc/crsttypes.h:89]: It is fine to add new entries to CrstTypes.def and regenerate CrstTypes.h.

Comment (UnityAlex) [reply src/coreclr/inc/crsttypes.h:89]: Great thanks!

File: src/coreclr/vm/mono/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+if(WIN32)
+  add_definitions(-DPLATFORM_WIN=1)
+endif()
+
+if(CLR_CMAKE_PLATFORM_DARWIN)
+  add_definitions(-DPLATFORM_OSX=1)
+endif()
+
+add_library_clr(mono ${VM_SOURCES_MONO})
+#target_compile_features(mono PRIVATE cxx_range_for)
```
Comment (joshpeterson) [inline src/coreclr/vm/mono/CMakeLists.txt]: Do we still need this commented-out line?

**Review by joshpeterson — APPROVED** [review]

**Review by scott-ferguson-unity — APPROVED** [review]

**Review by UnityAlex — APPROVED** [review]

File: src/coreclr/vm/mono/CMakeLists.txt (hunk 4 above)
Comment (joncham) [reply src/coreclr/vm/mono/CMakeLists.txt]: Removed

**Pull request merged**